src/jdk.compiler/share/classes/com/sun/tools/javac/jvm/ClassReader.java

```diff
@@ -561,31 +540,46 @@ Type classSigToType() {
                                Convert.utf2string(signature, sigp, 10));
         sigp++;
         Type outer = Type.noType;
+        Name name;
+        boolean requireProjection;
         int startSbp = sbp;
 
         while (true) {
             final byte c = signature[sigp++];
             switch (c) {
 
             case ';': {         // end
-                ClassSymbol t = enterClass(names.fromUtf(signatureBuffer,
-                                                         startSbp,
-                                                         sbp - startSbp));
-
+                name = names.fromUtf(signatureBuffer,
```

Comment:
This is another bit which might get simpler if we end up generating only a single .class.